Data/ByteString/Internal.hs

```diff
@@ -460,24 +463,27 @@ createUptoN' l f = do
 --
 createAndTrim :: Int -> (Ptr Word8 -> IO Int) -> IO ByteString
 createAndTrim l f = do
-    fp <- mallocByteString l
-    withForeignPtr fp $ \p -> do
-        l' <- f p
-        if assert (l' <= l) $ l' >= l
-            then return $! PS fp 0 l
-            else create l' $ \p' -> memcpy p' p l'
+    p  <- mallocBytes l
```

Comment:
I'm not sure that we want to do this. The `mallocByteString` function uses `mallocPlainForeignPtrBytes` which historically has been quicker than using malloc + ForeignPtr (ForeignPtr has special support for this type).

@simonmar, @donsbot, @dcoutts can comment I think